### PR TITLE
Reduce Route memory utilization by avoiding Metadata instances when not needed

### PR DESCRIPTION
**Pull request by yurykats**
Reduce Route memory utilization by avoiding Metadata instances when not needed

Currently an instance of raw Metadata and an instance of TypedMetadata are created in each route, even if no metadata is configured. This PR switches instances to unique_ptr and only creates Metadata objects when actually configured. Otherwise a singleton default metadata is used.

This is a step towards more memory efficient config data structures. Issue #24154.

Signed-off-by: Yury Kats <ykats@google.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/router/config_impl.h [hunk 1]
```diff
@@ -1061,8 +1075,8 @@ class RouteEntryImplBase : public RouteEntryAndRoute,
   HeaderParserPtr request_headers_parser_;
   HeaderParserPtr response_headers_parser_;
   uint32_t retry_shadow_buffer_limit_{std::numeric_limits<uint32_t>::max()};
-  envoy::config::core::v3::Metadata metadata_;
-  Envoy::Config::TypedMetadataImpl<HttpRouteTypedMetadataFactory> typed_metadata_;
+  std::unique_ptr<envoy::config::core::v3::Metadata> metadata_;
+  std::unique_ptr<RouteTypedMetadata> typed_metadata_;
```
Comment (adisuissa) [inline source/common/router/config_impl.h:1079]: nit: should these be const?

File: source/common/router/config_impl.cc [hunk 2]
```diff
@@ -534,6 +533,10 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost,
                           validator),
       route_name_(route.name()), time_source_(factory_context.mainThreadDispatcher().timeSource()),
       random_value_header_name_(route.route().weighted_clusters().header_name()) {
+  if (route.has_metadata()) {
+    metadata_ = std::make_unique<envoy::config::core::v3::Metadata>(route.metadata());
+    typed_metadata_ = std::make_unique<RouteTypedMetadata>(route.metadata());
+  }
```
Comment (adisuissa) [inline source/common/router/config_impl.cc:539]: I find it interesting that both wrap the same metadata field.
I guess this is intentional (after reading this [comment thread](https://github.com/envoyproxy/envoy/pull/4723/files#r225213097)), but just wondering whether there should be some ordering between these, or at least both should not store the same metadata.

File: source/common/router/config_impl.h [hunk 3]
```diff
@@ -658,8 +658,22 @@ class RouteEntryImplBase : public RouteEntryAndRoute,
...
+      return *metadata_;
+    }
+    return DefaultMetadata::get();
+  }
+  using RouteTypedMetadata = Envoy::Config::TypedMetadataImpl<HttpRouteTypedMetadataFactory>;
+  const Envoy::Config::TypedMetadata& typedMetadata() const override {
+    if (typed_metadata_ != nullptr) {
+      return *typed_metadata_;
+    }
+    static RouteTypedMetadata* defaultTypedMetadata =
```
Comment (adisuissa) [inline source/common/router/config_impl.h:673]: nit: const.

Comment (adisuissa) [reply source/common/router/config_impl.h:673]: On a second thought, could this be a ConstSingleton by itself (i.e., not dependent on `DefaultMetadata::get()`)?
I'm trading off another object of empty Metadata object against dependencies between singletons. WDYT?

**Review by adisuissa — COMMENTED** [review]
Thanks, left a few minor comments.

File: source/common/router/config_impl.h (hunk 3 above)
Comment (pradeepcrao) [reply source/common/router/config_impl.h:673]: ConstSingleton as defined does not allow constructors that need input arguments.

File: source/common/router/config_impl.cc (hunk 2 above)
Comment (yurykats) [reply source/common/router/config_impl.cc:539]: Yes, I find it strange as well, but it does keep the same metadata in two forms -- as a raw protobuf, and converted to "typed" data structure. I am not aiming to change this here, just avoid creation of both when there is no actual metadata.
To make things even more confusing, RDS keeps a full copy of the entire raw protobuf anyway, so a raw copy of the metadata message here is pure overhead. Again, not for this PR.

File: source/common/router/config_impl.h (hunk 3 above)
Comment (yurykats) [reply source/common/router/config_impl.h:673]: What @pradeepcrao said, typed metadata needs a raw metadata instance to be created.

Comment (pradeepcrao) [reply source/common/router/config_impl.h:673]: Might be possible to do using type traits which will make this cleaner. I can attempt that in a follow-up PR.

File: source/common/router/config_impl.h (hunk 1 above)
Comment (yurykats) [reply source/common/router/config_impl.h]: Done

File: source/common/router/config_impl.h (hunk 3 above)
Comment (yurykats) [reply source/common/router/config_impl.h]: Added const. Leaving the singleton as is.

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!

**Review by zuercher — APPROVED** [review]

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
Note: no changes since last approval by @zuercher (just merge-main conflict that was solved), so this PR should be ok to merge.

**Pull request merged**